### PR TITLE
Implemented PXB-2657 - KMIP key management component 

### PR DESCRIPTION
**Pull request by altmannmarcelo**
https://jira.percona.com/browse/PXB-2657

Implemented Percona Server KMIP keyring component support.
Adjusted xtrabackup component to be more generic for further keyring
components. Now all parse is done at create_component_config_data.

Deprecated --stats functionality for the component keyring.

### CODE REVIEW

**Comment by altmannmarcelo** [comment]
@satya-bodapati @rahulmalik87 please review 86941fe commit only. The previous ones will be cherry-picked once again when PS KMIP work is completed.

File: components/keyrings/common/config/config_reader.h [hunk 1]
```diff
@@ -43,7 +43,19 @@ class Config_reader {
...
+    Constructor
+
+    Reads JSON from parameter and stores it in memory.
 
+    @param [in] config_file_path Full path to configuration file - not used
+
+    @param [in] config_data JSON string containing config data
+  */
+  explicit Config_reader(const std::string config_file_path,
+                         const std::string config_data);
```
Comment (satya-bodapati) [inline components/keyrings/common/config/config_reader.h]: use  reference here

File: storage/innobase/xtrabackup/src/CMakeLists.txt [hunk 2]
```diff
@@ -134,6 +134,11 @@ TARGET_LINK_LIBRARIES(xtrabackup
    ADD_INSTALL_RPATH_FOR_PROTOBUF(xtrabackup)
  ENDIF()
 
+ # KMIP depend on boost
+ INCLUDE_DIRECTORIES(SYSTEM
+        ${BOOST_PATCHES_DIR}
+        ${BOOST_INCLUDE_DIR})
+
```
Comment (satya-bodapati) [inline storage/innobase/xtrabackup/src/CMakeLists.txt:141]: will this link boost to plugin or xtrabackup?

File: storage/innobase/xtrabackup/test/inc/keyring_kmip.sh [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+fi
+
+if test -d $PWD/../../../../plugin_output_directory
+then
+  plugin_dir=$PWD/../../../../plugin_output_directory
+else
+  plugin_dir=$PWD/../../lib/plugin/
+fi
+XB_EXTRA_MY_CNF_OPTS="${XB_EXTRA_MY_CNF_OPTS:-""}
+xtrabackup-plugin-dir=${plugin_dir}
```
Comment (satya-bodapati) [inline storage/innobase/xtrabackup/test/inc/keyring_kmip.sh]: move this to common.inc?

File: storage/innobase/xtrabackup/test/inc/keyring_kmip.sh [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+"
+
+#placeholder
+keyring_args=""
+
+KMIP_SERVER_ADDR="${KMIP_SERVER_ADDR:-127.0.0.1}"
+KMIP_SERVER_PORT="${KMIP_SERVER_PORT:-5696}"
+KMIP_CLIENT_CA="${KMIP_CLIENT_CA:-/etc/pykmip/client_certificate_jane_doe.pem}"
+KMIP_CLIENT_KEY="${KMIP_CLIENT_KEY:-/etc/pykmip/client_key_jane_doe.pem}"
+KMIP_SERVER_CA="${KMIP_SERVER_CA:-/etc/pykmip/root_certificate.pem}"
```
Comment (satya-bodapati) [inline storage/innobase/xtrabackup/test/inc/keyring_kmip.sh:23]: how to get these files? can you please add a comment above?

File: storage/innobase/xtrabackup/test/suites/keyring/innodb_keyring_file_component.sh [hunk 5]
```diff
@@ -64,8 +64,6 @@ run_cmd_expect_failure $XB_BIN $XB_ARGS \
 --prepare --target-dir=$topdir/backup1 --xtrabackup-plugin-dir=${plugin_dir} \
 --component-keyring-file-config=$topdir/xtrabackup_component_keyring_file_invalid.cnf 2>&1 | tee $topdir/pxb.log
 
-grep_count "Component configuration does not have path member" $topdir/pxb.log 1
-
```
Comment (satya-bodapati) [inline storage/innobase/xtrabackup/test/suites/keyring/innodb_keyring_file_component.sh:68]: removed by mistake?

File: components/keyrings/common/config/config_reader.h [hunk 6]
```diff
@@ -43,7 +43,19 @@ class Config_reader {
     @param [in] config_file_path Full path to configuration file
   */
   explicit Config_reader(const std::string config_file_path);
+#ifdef XTRABACKUP
+  /**
+    Constructor
+
+    Reads JSON from parameter and stores it in memory.
 
+    @param [in] config_file_path Full path to configuration file - not used
```
Comment (satya-bodapati) [inline components/keyrings/common/config/config_reader.h]: Not used?  incase of offline (--prepare), we pass the file here? :thinking:

**Review by satya-bodapati — CHANGES_REQUESTED** [review]
Looks good. some minor comments.

File: components/keyrings/common/config/config_reader.h (hunk 6 above)
Comment (altmannmarcelo) [reply components/keyrings/common/config/config_reader.h]: We pass the config_data already populated,  from keyring config.cc - https://github.com/altmannmarcelo/percona-xtrabackup/blob/PXB-2657-8.0/components/keyrings/keyring_file/config/config.cc#L90, here we are just overloading the constructor to not try to load the manigest config file as they won't be on their set location  - https://dev.mysql.com/doc/refman/8.0/en/keyring-component-installation.html .

Problem here is that config_data and config_file_path are both of the same type, I might be able to adjust the Constructor to accept `rapidjson::StringBuffer` so we can overload it with a single parameter.

File: components/keyrings/common/config/config_reader.h (hunk 1 above)
Comment (altmannmarcelo) [reply components/keyrings/common/config/config_reader.h]: Done.

File: storage/innobase/xtrabackup/src/CMakeLists.txt (hunk 2 above)
Comment (altmannmarcelo) [reply storage/innobase/xtrabackup/src/CMakeLists.txt:141]:
```
marcelo@marce-bld:/work/pxb/src/8.0$ ldd /work/pxb/ins/8.0/bin/xtrabackup  | grep boost | wc -l
0
marcelo@marce-bld:/work/pxb/src/8.0$ ldd /work/pxb/ins/8.0/lib/plugin/component_keyring_kmip.so  | grep boost | wc -l
0
```
neither one.

File: storage/innobase/xtrabackup/test/inc/keyring_kmip.sh (hunk 4 above)
Comment (altmannmarcelo) [reply storage/innobase/xtrabackup/test/inc/keyring_kmip.sh:23]: Done

File: storage/innobase/xtrabackup/test/suites/keyring/innodb_keyring_file_component.sh (hunk 5 above)
Comment (altmannmarcelo) [reply storage/innobase/xtrabackup/test/suites/keyring/innodb_keyring_file_component.sh:68]: No, we no longer print this message at https://github.com/percona/percona-xtrabackup/pull/1226/commits/86941fe7c6a55b509a978fea15ba9c3c4b334a95#diff-d9a0076502e9950744186c9fe95e006525d8f050b1ee5828fcfe7d3e45134a82L217

File: storage/innobase/xtrabackup/test/inc/keyring_kmip.sh (hunk 3 above)
Comment (altmannmarcelo) [reply storage/innobase/xtrabackup/test/inc/keyring_kmip.sh]: Done

File: components/keyrings/common/config/config_reader.h (hunk 6 above)
Comment (altmannmarcelo) [reply components/keyrings/common/config/config_reader.h]: Done

**Review by satya-bodapati — APPROVED** [review]
LGTM!!!

**Comment by altmannmarcelo** [comment]
Local test results of `keyring_kmip`, `keyring_vault` and `keyring_file` :
```
KMIP_CLIENT_CA=/etc/pykmip/certs/client_certificate_jane_doe.pem KMIP_CLIENT_KEY=/etc/pykmip/certs/client_key_jane_doe.pem KMIP_SERVER_CA=/etc/pykmip/certs/root_certificate.pem VAULT_URL=https://192.168.23.44:8200 VAULT_CACERT=${HOME}/vault.crt VAULT_CA=${HOME}/vault.crt VAULT_TOKEN=s.wyUdaNYepSzySmYUDTBsmoTi ./run.sh -s keyring -k
Using /work/pxb/ins/8.0/xtrabackup-test/test/var as test root
Detecting server version...
Running against Percona Server 8.0.27-18-debug (XtraDB 8.0.27-18)
Using 'xtrabackup' as xtrabackup binary
Autodetected number of cores: 32
Limiting to 16 to avoid excessive resource consumption
Using 16 parallel workers
Running suite(s): keyring

==============================================================================
TEST                                   WORKER    RESULT     TIME(s) or COMMENT
------------------------------------------------------------------------------
bug1590351                               w1	[passed]    31
innodb_keyring_proxy_plugin              w10	[passed]    37
innodb_encryption_mix_plugins            w4	[passed]    49
keyring_encryption                       w16	[passed]    63
innodb_encryption_export                 w3	[passed]    75
pxb-1793                                 w16	[passed]    100
pxb-1705                                 w4	[passed]    123
innodb_redo_log_encrypt                  w15	[passed]    179
page_compression                         w10	[passed]    173
pxb-1903                                 w16	[passed]    56
keyring_pxb_2275                         w1	[passed]    203
pxb-1879                                 w3	[passed]    163
pxb-1937                                 w15	[passed]    67
pxb-1936                                 w4	[passed]    106
pxb-1954                                 w10	[passed]    75
pxb-2429                                 w16	[passed]    93
innodb_keyring_plugin_to_component       w9	[passed]    329
innodb_keyring_vault_auto_v1             w11	[passed]    330
innodb_keyring_vault_auto_v2             w12	[passed]    330
innodb_keyring_vault_v1                  w13	[passed]    330
innodb_keyring_vault_v2                  w14	[passed]    346
shared_tablespace_encryption             w3	[passed]    116
reencrypt                                w1	[passed]    131
bug1737525                               w2	[passed]    369
system_tablespace_encryption             w15	[passed]    134
xb_log_archiving_encrypt                 w4	[passed]    128
innodb_keyring_file_plugin               w7	[passed]    446
innodb_keyring_file_both                 w5	[passed]    465
innodb_keyring_file_component            w6	[passed]    496
innodb_keyring_kmip_component            w8	[passed]    500
==============================================================================
Spent 6043 of 515 seconds executing testcases

SUMMARY: 30 run, 30 successful, 0 skipped, 0 failed

See results/ for detailed output
```

**Pull request merged**